shared/trust/hmac.go

```diff
@@ -0,0 +1,265 @@
+}
+
+// ParseHeader extracts the actual version and HMAC from the header.
+func (h *HMACFormatDefault) ParseHeader(header string) (HMACVersion, []byte, error) {
+	authHeaderSplit := strings.Split(header, " ")
+	if len(authHeaderSplit) != 2 {
+		return "", nil, errors.New("Version or HMAC is missing")
+	}
+
+	hmac, err := hex.DecodeString(authHeaderSplit[1])
+	if err != nil {
+		return "", nil, fmt.Errorf("Failed to decode the HMAC: %w", err)
+	}
+
+	return HMACVersion(authHeaderSplit[0]), hmac, nil
+}
+
+// NewDefaultHMACFormatArgon returns the default format for argon2.
+// Recommended defaults according to https://www.rfc-editor.org/rfc/rfc9106#section-4-6.2.
+// We use the second recommended option to not require a system having 2 GiB of memory.
+func NewDefaultHMACFormatArgon() (*HMACFormatArgon, error) {
+	// 128 bit salt.
+	salt := make([]byte, 16)
+	_, err := rand.Read(salt)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create salt: %w", err)
+	}
+
+	return &HMACFormatArgon{
+		Salt: salt,
+		// 3 iterations.
+		Time: 3,
+		// 64 MiB memory.
+		Memory: 64 * 1024,
+		// 4 lanes.
+		Threads: 4,
+		// 256 bit tag size.
+		KeyLength: 32,
+	}, nil
+}
+
+// Passphrase accepts the outer HMAC passphrase and runs it through the argon key derivation function.
+func (h *HMACFormatArgon) Passphrase(passphrase []byte) []byte {
+	return argon2.IDKey(passphrase, h.Salt, h.Time, h.Memory, h.Threads, h.KeyLength)
+}
+
+// Header returns the actual HMAC alongside it's salt together with the used version.
+func (h *HMACFormatArgon) Header(version HMACVersion, hmac []byte) string {
+	return fmt.Sprintf("%s %s:%s", version, hex.EncodeToString(h.Salt), hex.EncodeToString(hmac))
+}
+
+// ParseHeader extracts the actual version, HMAC and it's salt from the header.
+func (h *HMACFormatArgon) ParseHeader(header string) (HMACVersion, []byte, error) {
+	authHeaderSplit := strings.Split(header, " ")
+	if len(authHeaderSplit) != 2 {
+		return "", nil, errors.New("Version or HMAC and salt combination is missing")
+	}
+
+	authHeaderDetails := strings.Split(authHeaderSplit[1], ":")
+	if len(authHeaderDetails) != 2 {
+		return "", nil, errors.New("Salt or HMAC is missing")
+	}
+
+	salt, err := hex.DecodeString(authHeaderDetails[0])
+	if err != nil {
+		return "", nil, fmt.Errorf("Failed to decode the salt: %w", err)
+	}
+
+	h.Salt = salt
+
+	hmac, err := hex.DecodeString(authHeaderDetails[1])
+	if err != nil {
+		return "", nil, fmt.Errorf("Failed to decode the HMAC: %w", err)
+	}
+
+	return HMACVersion(authHeaderSplit[0]), hmac, nil
+}
+
+// NewHMAC returns a new instance of HMAC.
+func NewHMAC(conf HMACConf) *HMAC {
+	return &HMAC{
+		conf:   conf,
+		format: &HMACFormatDefault{},
+	}
+}
+
+// WithFormat configures the HMAC instance to use the specified format.
+// The format can be used to extend the way of displaying and parsing the resulting header.
+func (h *HMAC) WithFormat(format HMACFormat) *HMAC {
```

Comment:
Can we just have different factory functions for the different formats, rather than having the ability to switch between them. I'm finding the method chaining functionality here a bit difficult to comprehend why it is needed?

Reply:
Fixed. There is now a `NewHMAC` and `NewHMACArgon2` which both return a struct that implements the `HMACFormatter`.